app/src/main/kotlin/com/simplemobiletools/smsmessenger/helpers/MessagesImporter.kt

```diff
@@ -1,66 +1,125 @@
     private var messagesFailed = 0
+    private var errorPassword : Boolean = false
+
+    private fun readBuffer(inputStream: InputStream, buffer: ByteArray)
+    {
+        var left = buffer.size
+
+        while(left > 0)
+        {
+            val count = inputStream.read(buffer, buffer.size - left, left)
+            if (count < 0)
+                throw Exception("Bad file encryption")
+            left -= count
+        }
+    }
 
-    fun importMessages(path: String, onProgress: (total: Int, current: Int) -> Unit = { _, _ -> }, callback: (result: ImportResult) -> Unit) {
+    fun importMessages(path: String, onProgress: (state: ImportState, total: Int, current: Int) -> Unit = { _, _, _ -> }, callback: (result: ImportResult) -> Unit) {
         ensureBackgroundThread {
             try {
-
                 val inputStream = if (path.contains("/")) {
                     File(path).inputStream()
                 } else {
                     context.assets.open(path)
                 }
 
-                inputStream.bufferedReader().use { reader ->
-                    val json = reader.readText()
-                    val type = object : TypeToken<List<ExportedMessage>>() {}.type
-                    val messages = gson.fromJson<List<ExportedMessage>>(json, type)
-                    val totalMessages = messages.flatMap { it.sms ?: emptyList() }.size + messages.flatMap { it.mms ?: emptyList() }.size
-                    if (totalMessages <= 0) {
-                        callback.invoke(IMPORT_NOTHING_NEW)
-                        return@ensureBackgroundThread
-                    }
+                val password = config.importBackupPassword
+
+                val data: InputStream
+                var rawInputStream: InputStream? = null
+
+                if (password != "")
+                {
+                    try {
+                        onProgress.invoke(ImportState.DECRYPTING, 0, 0)
+                        rawInputStream = BufferedInputStream(inputStream)
+                        val salt = ByteArray(16)
+                        val prefix = ByteArray(16)
+                        readBuffer(rawInputStream, salt)
+                        readBuffer(rawInputStream, prefix)
+                        val keyFactory: SecretKeyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1")
+                        val keySpec: KeySpec = PBEKeySpec(password.toCharArray(), salt, KEY_ITERATIONS, KEY_LENGTH)
+                        val secret: SecretKey = keyFactory.generateSecret(keySpec)
+                        val cipher: Cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
+                        val iv = IvParameterSpec(prefix)
+                        cipher.init(Cipher.DECRYPT_MODE, secret, iv)
+                        data = CipherInputStream(rawInputStream, cipher)
+
+                    } catch (e: Exception) {
+                        throw Exception("Error while decrypting backup, please check your password.")
```

Comment:
show some reasonable error messages to users on exceptions, dont crash the app

Reply:
The exception is thrown within the main try{}, it shouldn't crash the app (but rather show a toast), if I'm not mistaken.

Reply:
Speaking of this, it seems that there is really 2 state : IMPORT_FAIL or IMPORT_OK. IMPORT_PARTIAL _can_ be true, but it could be multiple other issues (file deleted, etc) maybe we can delete this and only show error, or success ?